### PR TITLE
[Intel MKL] Adding support for MKL to docker CI infrastructure.

### PR DESCRIPTION
**Pull request by claynerobison**
This PR adds support for the MKL to the docker CI infrastructure. Non-development containers are supported via the publicly available pip whls at https://anaconda.org/intel/tensorflow/. Development containers are supported as well. 

This PR also adds support for the following parameters:

- CI_DOCKER_EXTRA_BUILD_PARAMS: allows callers to pass parameters to the docker build command (e.g. --build-arg)
- TF_BAZEL_BUILD_OPTIONS: allows callers to pass parameters to the bazel build command (e.g. "--copt='mavx2'")

The MKL Dockerfiles have been renamed per the convention in parameterized_docker_build.sh. They have also been modified to work better with the sed commands in the build scripts.

### CODE REVIEW

File: tensorflow/tools/docker/Dockerfile.mkl
Comment (inc0) [inline tensorflow/tools/docker/Dockerfile.mkl]: aren't you maintainer Clayne?

File: tensorflow/tools/docker/parameterized_docker_build.sh
Comment (inc0) [inline tensorflow/tools/docker/parameterized_docker_build.sh]: you've introduced this env (and mechanism for it to work), but I can't see any place where you actually use it? Why?

**Review by inc0 — CHANGES_REQUESTED** [review]

File: tensorflow/tools/docker/Dockerfile.mkl
Comment (claynerobison) [reply tensorflow/tools/docker/Dockerfile.mkl]: oh yeah.

File: tensorflow/tools/docker/parameterized_docker_build.sh
Comment (claynerobison) [reply tensorflow/tools/docker/parameterized_docker_build.sh]: See line 332.

Comment (inc0) [reply tensorflow/tools/docker/parameterized_docker_build.sh]: yeah you allowed to set TF_BAZEL_BUILD_OPTIONS and provided mechanism for it to work. I assume what will actually use this mechanism is part of CI somewhere? Not in this change?

Comment (claynerobison) [reply tensorflow/tools/docker/parameterized_docker_build.sh]: Correct. We have scripts internally that use it. tensorflow/tools/ci_build/linux/cpu/run_mkl.sh could (and probably should) use it. Right now, that script just takes default compiler x86 optimizations. tensorflow/tools/ci_build/ci_parameterized_build.sh has a TF_BUILD_OPTIONS parameter that is close, but doesn't give our scripts the flexibility they need.

Comment (inc0) [reply tensorflow/tools/docker/parameterized_docker_build.sh]: OK, makes sense. Normally I'd say we should do it in separate PR as this is different feature/problem, but that's ok. Smaller PRs are just easier to review

**Review by inc0 — APPROVED** [review]

**Comment by rmlarsen** [comment]
@jhseu could you take a look, please?

**Comment by jhseu** [comment]
The changes look good to me, but I'm switching reviewers to @gunan since there are test infrastructure changes going on, so I don't know if this is impacted.

**Comment by gunan** [comment]
All of the bash files under here are on their way out. @case540 is working on removing those.
If you like to check in CI scripts, I recommend creating standalone scripts that do not need any flags under `tools/ci_build/linux/mkl`

**Comment by claynerobison** [comment]
@case540 Can you give me some idea of what you are woring on  so I can help add MKL support to it?

**Comment by case540** [comment]
Yeah, working on rewriting these bash scripts in python to make the code more organized/readable.  Anything you add to the current .sh scripts will probably be deleted (or reimplemented in Python). 

Not much you can currently do to help except by changing the current build shell scripts as little as possible.

**Comment by inc0** [comment]
@case540 how about we merge this one (since python are still in the making) and we'll rewrite shells to .py as soon as you're ready?

**Comment by case540** [comment]
That would be fine with me.

**Comment by rmlarsen** [comment]
@case540 could you please review and approve this, if you are OK with merging it?

File: tensorflow/tools/docker/parameterized_docker_build.sh
Comment (case540) [inline tensorflow/tools/docker/parameterized_docker_build.sh]: Not really familiar with Dockerfiles, but is there a cleaner/less brittle way of modifying the Bazel build options?

File: tensorflow/tools/docker/parameterized_docker_build.sh
Comment (case540) [inline tensorflow/tools/docker/parameterized_docker_build.sh]: Not a huge fan of using sed on the dockerfile like this. Is there a better way to add in mkl build params?

Comment (claynerobison) [reply tensorflow/tools/docker/parameterized_docker_build.sh]: I don't think it's terribly elegant either, but it is what the original authors did for gpu builds in the lines just before this, so I thought it best to follow the model.

File: tensorflow/tools/docker/parameterized_docker_build.sh
Comment (claynerobison) [reply tensorflow/tools/docker/parameterized_docker_build.sh]: Same comment as above. I'm just following the pattern that the original authors were using, regardless of elegance. 

Dockerfiles themselves can have variables that can be overridden on the docker build command. For example, if the following line is in a Dockerfile:

`# ARGs are parameterized. Use --build-args to override.
+ARG WHL_DIR=/whl`

then when building the container, the caller can override the default value of WHL_DIR like this:

`docker build --build-arg WHL_DIR=<some path> . -f <docker file>`

That would be more elegant in my opinion.

File: tensorflow/tools/docker/parameterized_docker_build.sh
Comment (case540) [reply tensorflow/tools/docker/parameterized_docker_build.sh]: I was hoping there was a cleaner way to write the dockerfile before merging this. Using "sed" on the very specific line in the docker file feels pretty brittle.

But, Im more than fine merging this before the python impl is ready.

**Comment by rmlarsen** [comment]
@case540 & @claynerobison what is the status here? Can we proceed?
@claynerobison would you please rebase?

**Comment by claynerobison** [comment]
Closing because the rebase was a disaster. See new PR #18745

**Pull request closed without merging**
